File: src/main/kotlin/leetcode/string/CustomSortString.java
package leetcode.string;

import java.util.Arrays;

public class CustomSortString {
    public static String customSortString(String order, String s) {
        Integer[] indexes = new Integer[s.length()];
        for (int i = 0; i < s.length(); i++) {
            indexes[i] = i;
        }
        
        Arrays.sort(indexes, (a, b) -> {
            int aPriority = order.indexOf(s.charAt(a));
            int bPriority = order.indexOf(s.charAt(b));
            if (aPriority == -1) aPriority = order.length();
            if (bPriority == -1) bPriority = order.length();
            return Integer.compare(aPriority, bPriority);
        });
        
        StringBuilder sb = new StringBuilder();
        for (int index : indexes) {
            sb.append(s.charAt(index));
        }
        return sb.toString();
    }

    public static void main(String[] args) {
        System.out.println("Example 1 Output: " + customSortString("cba", "abcd")); // Should print "cbad"
        System.out.println("Example 2 Output: " + customSortString("bcafg", "abcd")); // Should print "bcad"
    }
}

File: src/main/kotlin/leetcode/string/customSortString.kt
package leetcode.string


private fun main(){
    val order = "bcafg"
    val s = "abcd"
    customSortString(order,s)
}
fun customSortString(order: String, s: String): String {
    val priorityMap = order.withIndex().associate { it.value to it.index }
    return s.toCharArray().sortedWith(Comparator { a, b ->
        val aPriority = priorityMap.getOrDefault(a, order.length)
        val bPriority = priorityMap.getOrDefault(b, order.length)
        aPriority.compareTo(bPriority)
    }).joinToString("")
}

//97
//98
//99
//[a, b, c]

//97
//98
//99
//102
//103
//[a, b, c, f, g]
